connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -89,8 +90,29 @@ class SparkSession(
   @Experimental
   def sql(sqlText: String, args: java.util.Map[String, String]): DataFrame = newDataset {
     builder =>
-      builder
-        .setSql(proto.SQL.newBuilder().setQuery(sqlText).putAllArgs(args))
+      // Send the SQL once to the server and then check the output.
+      val cmd = newCommand(b =>
+        b.setSqlCommand(proto.SqlCommand.newBuilder().setSql(sqlText).putAllArgs(args)))
+      val plan = proto.Plan.newBuilder().setCommand(cmd)
+      val response = client.execute(plan.build())
+      val (data, isSqlCommand) =
+        response.asScala.foldLeft((Seq[ByteString](), false))((acc, resp) =>
+          if (resp.hasSqlCommandResult) {
+            (acc._1, resp.getSqlCommandResult.getIsCommand)
+          } else if (resp.hasArrowBatch) {
+            (acc._1 ++ Seq(resp.getArrowBatch.getData), acc._2)
+          } else {
+            acc
+          })
+
+      // Handle the result behavior
+      if (isSqlCommand) {
+        assert(data.size <= 1)
```

Comment:
We always send one batch right? Even if it is empty, you should always get the schema.